docker-compose-rule-core/src/main/java/com/palantir/docker/compose/DockerComposeManager.java

```diff
@@ -0,0 +1,380 @@
+
+        waitForAllClusterWaits(allClusterWaits);
+
+        log.debug("docker-compose cluster started");
+    }
+
+    private void waitForAllClusterWaits(List<InterruptableClusterWait> allClusterWaits) throws InterruptedException {
+        ListeningExecutorService executorService = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(
+                allClusterWaits.size(),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("dcr-wait-%d")
+                        .build()));
+
+        try {
+            ListenableFuture<?> listListenableFuture =
+                    Futures.allAsList(allClusterWaits.stream()
+                    .map(clusterWait -> executorService.submit(() -> {
+                        try {
+                            clusterWait.waitForCluster(containers());
+                        } catch (InterruptedException e) {
+                            if (executorService.isShutdown()) {
+                                // ignore if this InterruptedException has occurred because we shut down and
+                                // terminated the executor
+                                return;
+                            }
+
+                            Throwables.propagate(e);
+                        }
+                    }))
+                    .collect(Collectors.toList()));
+
+            listListenableFuture.get();
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof RuntimeException) {
+                throw (RuntimeException) e.getCause();
+            }
+            throw new IllegalStateException("A cluster wait errored out: ", e);
+        } finally {
+            MoreExecutors.shutdownAndAwaitTermination(executorService, 0, TimeUnit.SECONDS);
+        }
+    }
+
+    public void after() {
+        try {
+            emitEventsFor().shutdownStop(() ->
+                    shutdownStrategy().stop(this.dockerCompose()));
+
+            emitEventsFor().logCollection(() ->
+                    logCollector().collectLogs(this.dockerCompose()));
+
+            emitEventsFor().shutdown(() ->
+                    shutdownStrategy().shutdown(this.dockerCompose(), this.docker()));
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException("Error cleaning up docker compose cluster", e);
+        }
+    }
+
+    public String exec(DockerComposeExecOption options, String containerName,
+            DockerComposeExecArgument arguments) throws IOException, InterruptedException {
+        return dockerCompose().exec(options, containerName, arguments);
+    }
+
+    public String run(DockerComposeRunOption options, String containerName,
+            DockerComposeRunArgument arguments) throws IOException, InterruptedException {
+        return dockerCompose().run(options, containerName, arguments);
+    }
+
+    /*
+     *  TODO(forozco)
+     */
+    public static Builder testBuilder() {
```

Comment:
Why is this called `testBuilder` rather than `builder`?

Reply:
Sorry, I meant to flesh out the comment. To get `DockerComposeRule` and `DockerComposeExtension` working with as little duplicate code as possible, I had to make them both extend `DockerComposeManager`. Unfortunately, its not possible to get Immutable builders to extend from each other. So to get `DockerComposeRule.builder()` to worker we could not have a `builder()` method on `DockerComposeManager`